include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+   * @param data_ptr is a location to store the data dequeued.
+   */
+  virtual WasmResult dequeueSharedQueue(uint32_t /* token */, std::string * /* data_ptr */);
+
+  /**
+   * Enqueue a message on a shared queue.
+   * @param token is a token returned by resolveSharedQueue();
+   * @param data is the data to be queued.
+   */
+  virtual WasmResult enqueueSharedQueue(uint32_t /* token */, string_view /* value */);
+
+  // Header/Trailer/Metadata Maps
+
+  /**
+   * Add a key-value pair to a header map.
+   * @param type of the header map.
+   * @param key is the key (header).
+   * @param value is the value (header value).
+   */
+  virtual WasmResult addHeaderMapValue(WasmHeaderMapType /* type */, string_view /* key */,
+                                       string_view /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Get a value from to a header map.
+   * @param type of the header map.
+   * @param key is the key (header).
+   * @param result is a pointer to the returned header value.
+   */
+  virtual WasmResult getHeaderMapValue(WasmHeaderMapType /* type */, string_view /* key */,
+                                       string_view *result) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Get all the key-value pairs in a header map.
+   * @param type of the header map.
+   * @param result is a pointer to the pairs.
+   */
+  virtual WasmResult getHeaderMapPairs(WasmHeaderMapType /* type */, Pairs * /* result */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Set a header map so that it contains the given pairs (does not merge with existing data).
+   * @param type of the header map.
+   * @param the pairs to set the header map to.
+   */
+  virtual WasmResult setHeaderMapPairs(WasmHeaderMapType /* type */, const Pairs & /* pairs */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Remove a key-value pair from a header map.
+   * @param type of the header map.
+   * @param key of the header map.
+   */
+  virtual WasmResult removeHeaderMapValue(WasmHeaderMapType /* type */, string_view /* key */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Replace (or set) a value in a header map.
+   * @param type of the header map.
+   * @param key of the header map.
+   * @param value to set in the header map.
+   */
+  virtual WasmResult replaceHeaderMapValue(WasmHeaderMapType /* type */, string_view /* key */,
+                                           string_view /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Returns the number of entries in a header map.
+   * @param type of the header map.
+   * @param result is a pointer to the result.
+   */
+  virtual WasmResult getHeaderMapSize(WasmHeaderMapType /* type */, uint32_t * /* result */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
```

Comment:
The other alternative to ^^ is an iterator interface for header maps (like Envoy internal), did you consider that?

Reply:
Yes.  We chose this instead because the cost of moving data into the VM is smaller than the cost of traversing the VM boundary frequently.  Consequently for large scale header searching and manipulation with bounded size headers it is likely faster to just move the data into the VM and do the work there.  It is also more flexible because we can update the SDK quickly and in a language specific way without having to change the ABI version.